app/forms/mappings/LocalDateFormatter.scala

```diff
@@ -24,70 +24,55 @@ import scala.util.{Failure, Success, Try}
 
 private[mappings] class LocalDateFormatter(
   invalidKey: String,
-  allRequiredKey: String,
-  twoRequiredKey: String,
   requiredKey: String,
   args: Seq[String] = Seq.empty
 ) extends Formatter[LocalDate]
     with Formatters {
 
-  private val fieldKeys: List[String] = List("Day", "Month", "Year")
+  private val fieldKeys: List[String] = List("day", "month", "year")
 
   private def toDate(key: String, day: Int, month: Int, year: Int): Either[Seq[FormError], LocalDate] =
     Try(LocalDate.of(year, month, day)) match {
       case Success(date) =>
         Right(date)
       case Failure(_) =>
-        Left(Seq(FormError(key, invalidKey, args)))
+        Left(Seq(FormError(key, s"$invalidKey.all", fieldKeys)))
     }
 
-  private def formatDate(key: String, data: Map[String, String]): Either[Seq[FormError], LocalDate] = {
-
-    val int = intFormatter(
-      requiredKey = invalidKey,
-      wholeNumberKey = invalidKey,
-      nonNumericKey = invalidKey,
-      args
-    )
-
-    for {
-      day   <- int.bind(s"${key}Day".replaceAll("\\s", ""), data)
-      month <- int.bind(s"${key}Month".replaceAll("\\s", ""), data)
-      year  <- int.bind(s"${key}Year".replaceAll("\\s", ""), data)
-      date  <- toDate(key, day, month, year)
-    } yield date
-  }
-
-  override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], LocalDate] = {
-
-    val fields: Map[String, Option[String]] = fieldKeys.map {
-      field =>
-        field -> data.get(s"$key$field").filter(_.nonEmpty).map(_.replaceAll("\\s", ""))
-    }.toMap
-
-    lazy val missingFields = fields
-      .withFilter(_._2.isEmpty)
-      .map(_._1.toLowerCase)
-      .toList
-
-    fields.count(_._2.isDefined) match {
-      case 3 =>
-        formatDate(key, data).left.map {
-          _.map(_.copy(key = key, args = args))
+  override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], LocalDate] =
+    fieldKeys
+      .foldLeft[Seq[Either[FormError, Int]]](Nil) {
```

Comment:
Because there are only three elements in the listo `List("day", "month", "year"),` could we do it in a for comp, I think it may eliminate the error parsing too? (I think :s)

Reply:
@joecefnhedog a for comprehension wouldn't help us here as if the binding does return a Left for a particular field we need to establish the nature of what caused it to be a Left. We need to do that for each field, collate the results and determine:

1. how many fields were unpopulated
2. how many fields had invalid values (i.e. non-numerics)